src/common/src/array/list_array.rs

```diff
@@ -624,6 +626,197 @@ impl<'a> From<&'a ListValue> for ListRef<'a> {
     }
 }
 
+impl ListValue {
+    /// Construct an array from literal string.
+    pub fn from_str(input: &str, data_type: &DataType) -> Result<Self, String> {
+        struct Parser<'a> {
+            input: &'a str,
+            data_type: &'a DataType,
+        }
+
+        impl Parser<'_> {
+            /// Parse a datum.
+            fn parse(&mut self) -> Result<Datum, String> {
+                self.skip_whitespace();
+                if self.try_parse_null() {
+                    return Ok(None);
+                }
+                if self.data_type.is_array() {
+                    Ok(Some(self.parse_array()?.into()))
+                } else {
+                    self.parse_value()
```

Comment:
Given `null` is already handled above, `parse_value` (and `parse_quoted`) should be returning `ScalarImpl` rather than `Datum`.